src/main/java/org/elasticsearch/common/util/concurrent/ReleasableLock.java

```diff
@@ -30,17 +30,47 @@
 public class ReleasableLock implements Releasable {
     private final Lock lock;
 
+    private final ThreadLocal<Boolean> holdingThreads;
+
     public ReleasableLock(Lock lock) {
         this.lock = lock;
+        boolean useHoldingThreads = false;
+        assert (useHoldingThreads = true);
```

Comment:
Maybe you can leave a comment here about what the holdingThreads ThreadLocal is used for, that way it doesn't get accidentally changed during a cleanup at a later time.